feat(ivc): impl `IVC::new`

cyphersnake opened the pull request:
**Why?**
Part of #31. Realizes the zero step and uses the base case inside the
folding chip

**What?**

- Implementation of `IVC::new` method, where IVC is 'incrementally
  verifiable computation'.

- The `CommitmentKey::commit` now returns option to be able to handle an
  error (smaller key size than the input data size) on the function
  output, rather than panic internally

- Infrastructure for sha256 example for storing the commitment key in
  the local folder cache.

- Impl `assign_all_*` for `*CyclicAssignor` for easy assign collections
  & iterators.

**How?**

- Updated `sha256/main.rs`:
  - Implemented `get_or_create_commitment_key` function to manage
    commitment keys efficiently, either loading them from a file or
    generating new ones if not present.
  - Added logging for better tracking and debugging purposes.
  - Modified the main function to accommodate changes in commitment key
    handling and logging.

- Updated `commitment.rs`:
  - Enhanced `CommitmentKey` struct to handle errors more gracefully by
    returning `Option` instead of panicking.
  - Added functionality to save and load commitment keys to and from
    files, improving usability and persistence.

- In `incrementally_verifiable_computation.rs`, implemented `IVC::new`
  - `IVC::new` establishes the initial state of the IVC system, setting
    up the primary and secondary step circuits and configuring the
    necessary parameters.

- Updated `public_params.rs` to accommodate the new commitment key
  handling and structure changes in IVC and related components.

- Made various updates across other modules (`main_gate.rs`,
  `nifs/vanilla.rs`, `plonk/mod.rs`, `poseidon/random_oracle.rs`,
  `table.rs`) to support the new functionality and changes in the IVC
  system, including adjustments to handle commitment keys, error
  handling, and integration with the random oracle model.

cyphersnake commented:
Now the problem is somewhere inside `TableData`, asked @chaosma  for help 
```
thread 'main' panicked at /home/q99/projects/github.com/snarkify/sirius/src/polynomial/mod.rs:538:9:
assertion failed: self == other
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:127:5
   3: sirius::polynomial::Monomial<F>::add
             at ./src/polynomial/mod.rs:538:9
   4: sirius::polynomial::MultiPolynomial<F>::reduce
             at ./src/polynomial/mod.rs:700:17
   5: <sirius::polynomial::MultiPolynomial<F> as core::ops::arith::Mul>::mul
             at ./src/polynomial/mod.rs:758:9
   6: sirius::polynomial::Expression<F>::_expand::{{closure}}
             at ./src/polynomial/mod.rs:217:21
   7: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:160:17
   8: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
   9: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  10: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  11: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  12: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  13: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  14: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  15: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  16: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  17: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  18: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  19: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:163:25
  20: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  21: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:163:25
  22: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  23: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  24: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  25: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  26: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  27: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  28: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  29: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  30: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  31: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  32: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  33: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  34: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  35: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  36: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  37: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  38: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  39: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:159:25
  40: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  41: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:153:25
  42: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  43: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  44: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  45: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  46: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  47: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  48: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  49: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  50: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  51: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  52: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  53: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  54: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  55: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  56: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  57: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  58: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  59: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  60: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  61: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  62: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  63: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:158:25
  64: sirius::polynomial::Expression<F>::evaluate::{{closure}}
             at ./src/polynomial/mod.rs:142:13
  65: sirius::polynomial::Expression<F>::evaluate
             at ./src/polynomial/mod.rs:154:25
  66: sirius::polynomial::Expression<F>::_expand
             at ./src/polynomial/mod.rs:171:9
  67: sirius::polynomial::Expression<F>::expand
             at ./src/polynomial/mod.rs:227:9
  68: sirius::table::TableData<F>::plonk_structure
             at ./src/table.rs:245:20
  69: sirius::ivc::incrementally_verifiable_computation::IVC<_,_,C1,C2,SC1,SC2>::new
             at ./src/ivc/incrementally_verifiable_computation.rs:108:48
  70: sha256::main
             at ./examples/sha256/main.rs:395:19
  71: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

cyphersnake commented:
Fixed by #130 
New bug!
At https://github.com/snarkify/sirius/blob/5daebf67d9a306c74299af5af1bda355e1590526/src/table.rs#L272
After `primary.synth`

```
[2024-01-29T20:46:13Z DEBUG sirius::ivc::fold_relaxed_plonk_instance_chip] fold: challenges folded: []
thread 'main' panicked at /home/q99/projects/github.com/snarkify/sirius/src/table.rs:272:33:
called `Result::unwrap()` on an `Err` value: TooLongInput { input_len: 29360128, limit: 4194304 }
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
   4: sirius::table::TableData<F>::run_sps_protocol_0
             at ./src/table.rs:272:18
   5: sirius::table::TableData<F>::run_sps_protocol
             at ./src/table.rs:452:18
   6: sirius::ivc::incrementally_verifiable_computation::IVC<_,_,C1,C2,SC1,SC2>::new
             at ./src/ivc/incrementally_verifiable_computation.rs:201:67
   7: trivial::main
             at ./examples/trivial/main.rs:95:19
   8: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
@chaosma Actually you can help tomorrow morning, I'll join you as soon as I get to work. Note, to make things easier, there is now an example 'trivial' instead of 'sha256'

chaosma commented:
> Fixed by #130 New bug! At
> 
> https://github.com/snarkify/sirius/blob/5daebf67d9a306c74299af5af1bda355e1590526/src/table.rs#L272
> 
> 
> After `primary.synth`
> ```
> [2024-01-29T20:46:13Z DEBUG sirius::ivc::fold_relaxed_plonk_instance_chip] fold: challenges folded: []
> thread 'main' panicked at /home/q99/projects/github.com/snarkify/sirius/src/table.rs:272:33:
> called `Result::unwrap()` on an `Err` value: TooLongInput { input_len: 29360128, limit: 4194304 }
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
>    1: core::panicking::panic_fmt
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
>    2: core::result::unwrap_failed
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
>    3: core::result::Result<T,E>::unwrap
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
>    4: sirius::table::TableData<F>::run_sps_protocol_0
>              at ./src/table.rs:272:18
>    5: sirius::table::TableData<F>::run_sps_protocol
>              at ./src/table.rs:452:18
>    6: sirius::ivc::incrementally_verifiable_computation::IVC<_,_,C1,C2,SC1,SC2>::new
>              at ./src/ivc/incrementally_verifiable_computation.rs:201:67
>    7: trivial::main
>              at ./examples/trivial/main.rs:95:19
>    8: core::ops::function::FnOnce::call_once
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> ```
> 
> I can increase K to `24`, but I don't understand why it's happened
> 
> @chaosma Actually you can help tomorrow morning, I'll join you as soon as I get to work. Note, to make things easier, there is now an example 'trivial' instead of 'sha256'

I was working on snark-verifier this morning and early afternoon. Will start looking at this issue.

cyphersnake commented:
Fixed!
Separated the different types of `k` and renamed it to 
- `CIRCUIT_TABLE_SIZE`
- `COMMITMENT_KEY_SIZE`
With 22 & 25 it's works!

New bug:
For zero-round of secondary_circuit.synth (base case) we use plonk instance (`u`) from primary circuit (generate from table data after primary_circuit.synth) & now `PlonkInstance::W` points out of C2 curve & we can't assign them

Let's investigate

cyphersnake commented:
Problem solved!

The next thing on the debug agenda is the absence of instance columns inside `u` passed from the primary case
```rust
thread 'main' panicked at /home/mikhail/sirius/src/ivc/fold_relaxed_plonk_instance_chip.rs:832:74:
index out of bounds: the len is 0 but the index is 0
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::panic_bounds_check
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:190:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/slice/index.rs:258:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/slice/index.rs:18:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/vec/mod.rs:2732:9
   6: sirius::ivc::fold_relaxed_plonk_instance_chip::FoldRelaxedPlonkInstanceChip<_,C>::assign_current_relaxed
             at ./src/ivc/fold_relaxed_plonk_instance_chip.rs:832:74
   7: sirius::ivc::step_circuit::StepCircuitExt::synthesize_step_base_case::{{closure}}
             at ./src/ivc/step_circuit.rs:355:20
   8: <halo2_proofs::circuit::floor_planner::single_pass::SingleChipLayouter<F,CS> as halo2_proofs::circuit::Layouter<F>>::assign_region
             at /home/mikhail/.cargo/git/checkouts/halo2-6b4607134e1c93ad/98d3574/halo2_proofs/src/circuit/floor_planner/single_pass.rs:91:13
   9: sirius::ivc::step_circuit::StepCircuitExt::synthesize_step_base_case
             at ./src/ivc/step_circuit.rs:334:25
  10: sirius::ivc::step_circuit::StepCircuitExt::synthesize
             at ./src/ivc/step_circuit.rs:234:26
  11: sirius::ivc::incrementally_verifiable_computation::IVC<_,_,C1,C2,SC1,SC2>::new
             at ./src/ivc/incrementally_verifiable_computation.rs:263:17
  12: trivial::main
             at ./examples/trivial/main.rs:96:19
  13: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

cyphersnake commented:
Solved! 
![alive](https://github.com/snarkify/sirius/assets/57308842/ba1eaf59-8a1d-4262-982e-fb4d1e454448)

cyphersnake commented:
In general, we've been making changes to this branch together for the past few days, so @chaosma, you may not need me breaking it into smaller PRs for the review

However, I am fully prepared to do so and can break the changes into six batches tomorrow, be free to request it

cyphersnake commented:
Splited
